python/ray/experimental/state/state_cli.py

```diff
@@ -669,3 +677,321 @@ def object_summary(ctx, timeout: float, address: str):
+
+        ```
+        ray log --address=<ray address>
+        ```
+
+        [ray logs file] Print the last 500 lines of raylet.out on a head node.
+
+        ```
+        ray logs file raylet.out --tail 500
+        ```
+
+        [ray logs actor] Follow the log file with an actor id ABC.
+
+        ```
+        ray logs actor --id ABC --follow
+        ```
+    """
+    if ctx.invoked_subcommand is None:
+        # Forward to `ray logs file`
+        ctx.invoke(log_file)
+
+
+def _parse_log_node_target(
+    address: Optional[str], node_id: Optional[str], node_ip: Optional[str]
+):
+    """Parse the query API server and the target nodes from which logs should be retrieved
+
+    Args:
+        address: Address of the GCS of the head node running the API server. This is
+            usually set through RAY_ADDRESS, `auto`, or an explicit <ip:port>.
+        node_id: Id of the node containing the logs.
+        node_ip: Ip of the node containing the logs.
+    """
+    api_server_url = get_api_server_url(address)
+
+    if node_ip is None and node_id is None:
+        address = ray._private.services.canonicalize_bootstrap_address_or_die(address)
+        node_ip = address.split(":")[0]
+
+    return api_server_url, node_id, node_ip
+
+
+log_follow_option = click.option(
+    "--follow",
+    "-f",
+    required=False,
+    type=bool,
+    is_flag=True,
+    help="Streams the log file as it is updated instead of just tailing.",
+)
+
+log_tail_option = click.option(
+    "--tail",
+    required=False,
+    type=int,
+    default=DEFAULT_LOG_LIMIT,
+    help="Number of lines to tail from log. -1 indicates fetching the whole file.",
+)
+
+log_interval_option = click.option(
+    "--interval",
+    required=False,
+    type=float,
+    default=None,
+    help="The interval in secs to print new logs when `--follow` is specified.",
+    hidden=True,
+)
+
+log_timeout_option = click.option(
+    "--timeout",
+    default=DEFAULT_RPC_TIMEOUT,
+    help=(
+        "Timeout in seconds for the API requests. "
+        f"Default is {DEFAULT_RPC_TIMEOUT}. If --follow is specified, "
+        "this option will be ignored."
+    ),
+)
+
+log_node_ip_option = click.option(
+    "-ip",
+    "--node-ip",
+    required=False,
+    type=str,
+    default=None,
+    help="Filters the logs by this ip address",
+)
+
+log_node_id_option = click.option(
+    "--node-id",
+    "-id",
+    required=False,
+    type=str,
+    default=None,
+    help="Filters the logs by this NodeID",
+)
+
+
+def _print_log(
+    api_server_url: str = None,
+    node_id: Optional[str] = None,
+    node_ip: Optional[str] = None,
+    filename: Optional[str] = None,
+    actor_id: Optional[str] = None,
+    pid: Optional[int] = None,
+    follow: bool = False,
+    tail: int = DEFAULT_LOG_LIMIT,
+    timeout: int = DEFAULT_RPC_TIMEOUT,
+    interval: Optional[float] = None,
+):
+    """Wrapper around `get_log()` that prints the preamble and the log lines"""
```

Comment:
We are missing

```
        if not tail:
            tail = 0 if follow else DEFAULT_LOG_LIMIT
```
?

Reply:
Why would we want to set `tail=0` when `follow=True`?

`tail` will never be None with default values now.